server/src/main/java/org/elasticsearch/common/Strings.java

```diff
@@ -767,7 +771,13 @@ public static String toString(ToXContent toXContent, ToXContent.Params params) {
      * @param xContentBuilder builder containing an object to converted to a string
      */
     public static String toString(XContentBuilder xContentBuilder) {
-        return BytesReference.bytes(xContentBuilder).utf8ToString();
+        xContentBuilder.close();
+        OutputStream stream = xContentBuilder.getOutputStream();
+        if (stream instanceof ByteArrayOutputStream baos) {
+            return baos.toString(StandardCharsets.UTF_8);
+        } else {
+            return ((BytesStream) stream).bytes().utf8ToString();
```

Comment:
This seems fragile, there could be other output streams (eg wrappers/delegates/etc) in the future that could break this assumption?

Reply:
Yea for sure but as of today we make the same assumption in BytesReference.bytes(xContentBuilder) and I just leveraged that here to save some copying. Practically it's hard to imagine our codebase running into any other heap backed stream for the time being :)

Reply:
Maybe at least add an assert in the else case?

Reply:
I think I might have something better here that would also fix the weirdness we have to do around not closing the `BytesStream` in callers of this stuff.
Let me see if it works out, if not I'll open a follow-up to add the assert here and in the `BytesReference` code shortly :)